src/main/scala/vld/Validator.scala

```diff
@@ -0,0 +1,162 @@
+package vld
+
+/**
+  * Implement validator typeclass that should validate arbitrary value [T].
+  *
+  * @tparam T the type of the value to be validated.
+  */
+trait Validator[T] {
+  self =>
+  /**
+    * Validates the value.
+    *
+    * @param value value to be validated.
+    * @return Right(value) in case the value is valid, Left(message) on invalid value
+    */
+  def validate(value: T): Either[String, T]
+
+  /**
+    * And combinator.
+    *
+    * @param other validator to be combined with 'and' with this validator.
+    * @return the Right(value) only in case this validator and <code>other</code> validator returns valid value,
+    *         otherwise Left with error messages from both validators.
+    */
+  def and(other: Validator[T]): Validator[T] = {
+    new Validator[T] {
+      override def validate(t: T): Either[String, T] = {
+        self.validate(t) match {
+          case Left(m) => Left(m)
+          case Right(t) => other.validate(t)
+        }
+      }
+    }
+  }
+
+  /**
+    * Or combinator.
+    *
+    * @param other validator to be combined with 'or' with this validator.
+    * @return the Right(value) only in case either this validator or <code>other</code> validator returns valid value,
+    *         otherwise Left with error messages from the failed validator or from both if both failed.
+    */
+  def or(other: Validator[T]): Validator[T] = {
+    new Validator[T] {
+      override def validate(t: T): Either[String, T] = {
+        self.validate(t) match {
+          case Left(m1) => other.validate(t) match {
+            case Left(m2)=>Left(s"$m1 and $m2")
+            case Right(t)=>Right(t)
+          }
+          case Right(t) => Right(t)
+        }
+      }
+    }
+  }
+}
+
+
+object Validator {
+
+  implicit class PersonValidator(person: Person) {
+    def validate(v: Validator[Person]): Either[String, Person] = {
+      v.validate(person)
+    }
+  }
+
+  implicit class DefaultPersonValidator(person: Person) {
```

Comment:
there is no need to do DefaultXXX, and also no need to do it for every class. All you need to do is:
```scala
implicit class AnyValidator[T](t: T) {
    def validate(implicit v: Validator[T]): Either[String, T] = {
      v.validate(t)
    }
  }
```
then you either provide implicit validator in scope so you can just do `"".validate` or pass explicit validator `"".validate(nonEmpty)`

Reply:
Ok,thanks.